---
title: "Obtaining TCGA Data"
author: "Sasha Thalluri"
date: "8/23/2021"
output: html_document
---

This R Markdown will have the code to download the The Cancer Genome Atlas (TCGA) data used for subsequent analyses. In this document, we download masked CNV data from all of the 33 cancers available in the TCGA cohort. The output from this R Markdown will be a CSV with the masked CNV data for all 33 cancers that can be used for the downstream analyses. 

Masked CNV

```{r}
### Title: Getting all TCGA CNV Datasets using TGCAbiolinks
### Written by: Roshan Darji
### Date: 2018.09.27
### Last Updated: 2018.12.04
### Goal: Download TGCA KIRC hg38 datatypes and examine limiting data types.


# Initalizng Packages
library(TCGAbiolinks)
library(SummarizedExperiment)
library(DT)
library(httpuv)
library(devtools)

### 1. Getting valid project IDs using TGCAbiolinks and datatable
TCGAbiolinks::getGDCprojects()$project_id
TCGAbiolinks:::getProjectSummary("TCGA-KIRC")
datatable(TCGAbiolinks::getGDCprojects(),
          filter = "top",
          options = list(scrollX = T, keys = T, pagelength= 10),
          rownames = F,
          caption = "List of Projects")

##### Data #####
## TCGA-KIRC Masked Copy Number Variations
query.maskCNV.hg38.KIRC <- GDCquery(project = "TCGA-KIRC",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.KIRC$results[[1]]) # Total number of samples = Tumor + Normal = 1118
nrow(subset(query.maskCNV.hg38.KIRC$results[[1]], tissue.definition == "Solid Tissue Normal")) # 419 Normal samples
nrow(subset(query.maskCNV.hg38.KIRC$results[[1]], tissue.definition == "Primary solid Tumor")) # 586 Tumor samples
nrow(subset(query.maskCNV.hg38.KIRC$results[[1]], tissue.definition == "Blood Derived Normal")) # 112 Blood-derived Normal samples
# Missing 1
GDCdownload(query = query.maskCNV.hg38.KIRC)
maskCNV.KIRC <- GDCprepare(query = query.maskCNV.hg38.KIRC)
save(maskCNV.KIRC,
     file = "data/maskCNV.KIRC.RData")

## TCGA-BRCA Masked Copy Number Variations
query.maskCNV.hg38.BRCA <- GDCquery(project = "TCGA-BRCA",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.BRCA$results[[1]]) # Total number of samples = Tumor + Normal = 1118
nrow(subset(query.maskCNV.hg38.BRCA$results[[1]], tissue.definition == "Solid Tissue Normal")) # 419 Normal samples
nrow(subset(query.maskCNV.hg38.BRCA$results[[1]], tissue.definition == "Primary solid Tumor")) # 586 Tumor samples
nrow(subset(query.maskCNV.hg38.BRCA$results[[1]], tissue.definition == "Blood Derived Normal")) # 112 Blood-derived Normal samples
# Missing 1
GDCdownload(query = query.maskCNV.hg38.BRCA)
maskCNV.BRCA <- GDCprepare(query = query.maskCNV.hg38.BRCA)
save(maskCNV.BRCA,
     file = "data/maskCNV.BRCA.RData")

## TCGA-SARC Masked Copy Number Variations
query.maskCNV.hg38.SARC <- GDCquery(project = "TCGA-SARC",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.SARC$results[[1]]) # Total number of samples = Tumor + Normal = 522
nrow(subset(query.maskCNV.hg38.SARC$results[[1]], tissue.definition == "Solid Tissue Normal")) # 22 Normal samples
nrow(subset(query.maskCNV.hg38.SARC$results[[1]], tissue.definition == "Primary solid Tumor")) # 263 Tumor samples
nrow(subset(query.maskCNV.hg38.SARC$results[[1]], tissue.definition == "Blood Derived Normal")) # 233 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.SARC)
maskCNV.SARC <- GDCprepare(query = query.maskCNV.hg38.SARC)
save(maskCNV.SARC,
     file = "../data/maskCNV.SARC.RData")

## TCGA-MESO Masked Copy Number Variations
query.maskCNV.hg38.MESO <- GDCquery(project = "TCGA-MESO",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.MESO$results[[1]]) # Total number of samples = Tumor + Normal = 173
nrow(subset(query.maskCNV.hg38.MESO$results[[1]], tissue.definition == "Solid Tissue Normal")) # 1 Normal samples
nrow(subset(query.maskCNV.hg38.MESO$results[[1]], tissue.definition == "Primary solid Tumor")) # 87 Tumor samples
nrow(subset(query.maskCNV.hg38.MESO$results[[1]], tissue.definition == "Blood Derived Normal")) # 85 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.MESO)
maskCNV.MESO <- GDCprepare(query = query.maskCNV.hg38.MESO)
save(maskCNV.MESO,
     file = "../data/maskCNV.MESO.RData")

## TCGA-READ Masked Copy Number Variations
query.maskCNV.hg38.READ <- GDCquery(project = "TCGA-READ",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.READ$results[[1]]) # Total number of samples = Tumor + Normal = 317
nrow(subset(query.maskCNV.hg38.READ$results[[1]], tissue.definition == "Solid Tissue Normal")) # 17 Normal samples
nrow(subset(query.maskCNV.hg38.READ$results[[1]], tissue.definition == "Primary solid Tumor")) # 167 Tumor samples
nrow(subset(query.maskCNV.hg38.READ$results[[1]], tissue.definition == "Blood Derived Normal")) # 132 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.READ)
maskCNV.READ <- GDCprepare(query = query.maskCNV.hg38.READ)
save(maskCNV.READ,
     file = "../data/maskCNV.READ.RData")

## TCGA-KIRP Masked Copy Number Variations
query.maskCNV.hg38.KIRP <- GDCquery(project = "TCGA-KIRP",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.KIRP$results[[1]]) # Total number of samples = Tumor + Normal = 608
nrow(subset(query.maskCNV.hg38.KIRP$results[[1]], tissue.definition == "Solid Tissue Normal")) # 85 Normal samples
nrow(subset(query.maskCNV.hg38.KIRP$results[[1]], tissue.definition == "Primary solid Tumor")) # 304 Tumor samples
nrow(subset(query.maskCNV.hg38.KIRP$results[[1]], tissue.definition == "Blood Derived Normal")) # 219 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.KIRP)
maskCNV.KIRP <- GDCprepare(query = query.maskCNV.hg38.KIRP)
save(maskCNV.KIRP,
     file = "../data/maskCNV.KIRP.RData")

## TCGA-PAAD Masked Copy Number Variations
query.maskCNV.hg38.PAAD <- GDCquery(project = "TCGA-PAAD",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.PAAD$results[[1]]) # Total number of samples = Tumor + Normal = 368
nrow(subset(query.maskCNV.hg38.PAAD$results[[1]], tissue.definition == "Solid Tissue Normal")) # 37 Normal samples
nrow(subset(query.maskCNV.hg38.PAAD$results[[1]], tissue.definition == "Primary solid Tumor")) # 184 Tumor samples
nrow(subset(query.maskCNV.hg38.PAAD$results[[1]], tissue.definition == "Blood Derived Normal")) # 146 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.PAAD)
maskCNV.PAAD <- GDCprepare(query = query.maskCNV.hg38.PAAD)
save(maskCNV.PAAD,
     file = "../data/maskCNV.PAAD.RData")

## TCGA-GBM Masked Copy Number Variations
query.maskCNV.hg38.GBM <- GDCquery(project = "TCGA-GBM",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.GBM$results[[1]]) # Total number of samples = Tumor + Normal = 1149
nrow(subset(query.maskCNV.hg38.GBM$results[[1]], tissue.definition == "Solid Tissue Normal")) # 28 Normal samples
nrow(subset(query.maskCNV.hg38.GBM$results[[1]], tissue.definition == "Primary solid Tumor")) # 609 Tumor samples
nrow(subset(query.maskCNV.hg38.GBM$results[[1]], tissue.definition == "Blood Derived Normal")) # 499 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.GBM)
maskCNV.GBM <- GDCprepare(query = query.maskCNV.hg38.GBM)
save(maskCNV.GBM,
     file = "../data/maskCNV.GBM.RData")

## TCGA-ACC Masked Copy Number Variations
query.maskCNV.hg38.ACC <- GDCquery(project = "TCGA-ACC",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.ACC$results[[1]]) # Total number of samples = Tumor + Normal = 180
nrow(subset(query.maskCNV.hg38.ACC$results[[1]], tissue.definition == "Solid Tissue Normal")) # 5 Normal samples
nrow(subset(query.maskCNV.hg38.ACC$results[[1]], tissue.definition == "Primary solid Tumor")) # 90 Tumor samples
nrow(subset(query.maskCNV.hg38.ACC$results[[1]], tissue.definition == "Blood Derived Normal")) # 85 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.ACC)
maskCNV.ACC <- GDCprepare(query = query.maskCNV.hg38.ACC)
save(maskCNV.ACC,
     file = "../data/maskCNV.ACC.RData")

## TCGA-CESC Masked Copy Number Variations
query.maskCNV.hg38.CESC <- GDCquery(project = "TCGA-CESC",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.CESC$results[[1]]) # Total number of samples = Tumor + Normal = 586
nrow(subset(query.maskCNV.hg38.CESC$results[[1]], tissue.definition == "Solid Tissue Normal")) # 8 Normal samples
nrow(subset(query.maskCNV.hg38.CESC$results[[1]], tissue.definition == "Primary solid Tumor")) # 295 Tumor samples
nrow(subset(query.maskCNV.hg38.CESC$results[[1]], tissue.definition == "Blood Derived Normal")) # 281 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.CESC)
maskCNV.CESC <- GDCprepare(query = query.maskCNV.hg38.CESC)
save(maskCNV.CESC,
     file = "../data/maskCNV.CESC.RData")

## TCGA-ESCA Masked Copy Number Variations
query.maskCNV.hg38.ESCA <- GDCquery(project = "TCGA-ESCA",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.ESCA$results[[1]]) # Total number of samples = Tumor + Normal = 373
nrow(subset(query.maskCNV.hg38.ESCA$results[[1]], tissue.definition == "Solid Tissue Normal")) # 64 Normal samples
nrow(subset(query.maskCNV.hg38.ESCA$results[[1]], tissue.definition == "Primary solid Tumor")) # 184 Tumor samples
nrow(subset(query.maskCNV.hg38.ESCA$results[[1]], tissue.definition == "Blood Derived Normal")) # 124 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.ESCA)
maskCNV.ESCA <- GDCprepare(query = query.maskCNV.hg38.ESCA)
save(maskCNV.ESCA,
     file = "../data/maskCNV.ESCA.RData")

## TCGA-DLBC Masked Copy Number Variations
query.maskCNV.hg38.DLBC <- GDCquery(project = "TCGA-DLBC",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.DLBC$results[[1]]) # Total number of samples = Tumor + Normal = 94
nrow(subset(query.maskCNV.hg38.DLBC$results[[1]], tissue.definition == "Solid Tissue Normal")) # 0 Normal samples
nrow(subset(query.maskCNV.hg38.DLBC$results[[1]], tissue.definition == "Primary solid Tumor")) # 48 Tumor samples
nrow(subset(query.maskCNV.hg38.DLBC$results[[1]], tissue.definition == "Blood Derived Normal")) # 42 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.DLBC)
maskCNV.DLBC <- GDCprepare(query = query.maskCNV.hg38.DLBC)
save(maskCNV.DLBC,
     file = "../data/maskCNV.DLBC.RData")

## TCGA-KICH Masked Copy Number Variations
query.maskCNV.hg38.KICH <- GDCquery(project = "TCGA-KICH",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.KICH$results[[1]]) # Total number of samples = Tumor + Normal = 132
nrow(subset(query.maskCNV.hg38.KICH$results[[1]], tissue.definition == "Solid Tissue Normal")) # 57 Normal samples
nrow(subset(query.maskCNV.hg38.KICH$results[[1]], tissue.definition == "Primary solid Tumor")) # 66 Tumor samples
nrow(subset(query.maskCNV.hg38.KICH$results[[1]], tissue.definition == "Blood Derived Normal")) # 9 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.KICH)
maskCNV.KICH <- GDCprepare(query = query.maskCNV.hg38.KICH)
save(maskCNV.KICH,
     file = "../data/maskCNV.KICH.RData")

## TCGA-UVM Masked Copy Number Variations
query.maskCNV.hg38.UVM <- GDCquery(project = "TCGA-UVM",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.UVM$results[[1]]) # Total number of samples = Tumor + Normal = 160
nrow(subset(query.maskCNV.hg38.UVM$results[[1]], tissue.definition == "Solid Tissue Normal")) # 0 Normal samples
nrow(subset(query.maskCNV.hg38.UVM$results[[1]], tissue.definition == "Primary solid Tumor")) # 80 Tumor samples
nrow(subset(query.maskCNV.hg38.UVM$results[[1]], tissue.definition == "Blood Derived Normal")) # 80 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.UVM)
maskCNV.UVM <- GDCprepare(query = query.maskCNV.hg38.UVM)
save(maskCNV.UVM,
     file = "../data/maskCNV.UVM.RData")

## TCGA-LAML Masked Copy Number Variations
query.maskCNV.hg38.LAML <- GDCquery(project = "TCGA-LAML",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.LAML$results[[1]]) # Total number of samples = Tumor + Normal = 285
nrow(subset(query.maskCNV.hg38.LAML$results[[1]], tissue.definition == "Solid Tissue Normal")) # 142 Normal samples
nrow(subset(query.maskCNV.hg38.LAML$results[[1]], tissue.definition == "Primary solid Tumor")) # 0 Tumor samples
nrow(subset(query.maskCNV.hg38.LAML$results[[1]], tissue.definition == "Blood Derived Normal")) # 0 Blood-derived Normal samples
nrow(subset(query.maskCNV.hg38.LAML$results[[1]], tissue.definition == "Primary Blood Derived Cancer - Peripheral Blood")) # 143 Primary Blood-derived cancer samples

GDCdownload(query = query.maskCNV.hg38.LAML)
maskCNV.LAML <- GDCprepare(query = query.maskCNV.hg38.LAML)
save(maskCNV.LAML,
     file = "../data/maskCNV.LAML.RData")

## TCGA-SKCM Masked Copy Number Variations
query.maskCNV.hg38.SKCM <- GDCquery(project = "TCGA-SKCM",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.SKCM$results[[1]]) # Total number of samples = Tumor + Normal = 939
nrow(subset(query.maskCNV.hg38.SKCM$results[[1]], tissue.definition == "Solid Tissue Normal")) # 2 Normal samples
nrow(subset(query.maskCNV.hg38.SKCM$results[[1]], tissue.definition == "Primary solid Tumor")) # 104 Tumor samples
nrow(subset(query.maskCNV.hg38.SKCM$results[[1]], tissue.definition == "Blood Derived Normal")) # 465 Blood-derived Normal samples
nrow(subset(query.maskCNV.hg38.SKCM$results[[1]], tissue.definition == "Metastatic")) # 368 Metastatic samples

GDCdownload(query = query.maskCNV.hg38.SKCM)
maskCNV.SKCM <- GDCprepare(query = query.maskCNV.hg38.SKCM)
save(maskCNV.SKCM,
     file = "../data/maskCNV.SKCM.RData")

## TCGA-pcpg Masked Copy Number Variations
query.maskCNV.hg38.PCPG <- GDCquery(project = "TCGA-PCPG",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.PCPG$results[[1]]) # Total number of samples = Tumor + Normal = 365
nrow(subset(query.maskCNV.hg38.PCPG$results[[1]], tissue.definition == "Solid Tissue Normal")) # 5 Normal samples
nrow(subset(query.maskCNV.hg38.PCPG$results[[1]], tissue.definition == "Primary solid Tumor")) # 178 Tumor samples
nrow(subset(query.maskCNV.hg38.PCPG$results[[1]], tissue.definition == "Blood Derived Normal")) # 175 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.PCPG)
maskCNV.PCPG <- GDCprepare(query = query.maskCNV.hg38.PCPG)
save(maskCNV.PCPG,
     file = "../data/maskCNV.PCPG.RData")

## TCGA-COAD Masked Copy Number Variations
query.maskCNV.hg38.COAD <- GDCquery(project = "TCGA-COAD",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.COAD$results[[1]]) # Total number of samples = Tumor + Normal = 972
nrow(subset(query.maskCNV.hg38.COAD$results[[1]], tissue.definition == "Solid Tissue Normal")) # 90 Normal samples
nrow(subset(query.maskCNV.hg38.COAD$results[[1]], tissue.definition == "Primary solid Tumor")) # 502 Tumor samples
nrow(subset(query.maskCNV.hg38.COAD$results[[1]], tissue.definition == "Blood Derived Normal")) # 378 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.COAD)
maskCNV.COAD <- GDCprepare(query = query.maskCNV.hg38.COAD)
save(maskCNV.COAD,
     file = "../data/maskCNV.COAD.RData")

## TCGA-UCS Masked Copy Number Variations
query.maskCNV.hg38.UCS <- GDCquery(project = "TCGA-UCS",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.UCS$results[[1]]) # Total number of samples = Tumor + Normal = 110
nrow(subset(query.maskCNV.hg38.UCS$results[[1]], tissue.definition == "Solid Tissue Normal")) # 6 Normal samples
nrow(subset(query.maskCNV.hg38.UCS$results[[1]], tissue.definition == "Primary solid Tumor")) # 56 Tumor samples
nrow(subset(query.maskCNV.hg38.UCS$results[[1]], tissue.definition == "Blood Derived Normal")) # 48 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.UCS)
maskCNV.UCS <- GDCprepare(query = query.maskCNV.hg38.UCS)
save(maskCNV.UCS,
     file = "../data/maskCNV.UCS.RData")

## TCGA-LUSC Masked Copy Number Variations
query.maskCNV.hg38.LUSC <- GDCquery(project = "TCGA-LUSC",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.LUSC$results[[1]]) # Total number of samples = Tumor + Normal = 1059
nrow(subset(query.maskCNV.hg38.LUSC$results[[1]], tissue.definition == "Solid Tissue Normal")) # 239 Normal samples
nrow(subset(query.maskCNV.hg38.LUSC$results[[1]], tissue.definition == "Primary solid Tumor")) # 524 Tumor samples
nrow(subset(query.maskCNV.hg38.LUSC$results[[1]], tissue.definition == "Blood Derived Normal")) # 296 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.LUSC)
maskCNV.LUSC <- GDCprepare(query = query.maskCNV.hg38.LUSC)
save(maskCNV.LUSC,
     file = "../data/maskCNV.LUSC.RData")

## TCGA-LGG Masked Copy Number Variations
query.maskCNV.hg38.LGG <- GDCquery(project = "TCGA-LGG",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.LGG$results[[1]]) # Total number of samples = Tumor + Normal = 1019
nrow(subset(query.maskCNV.hg38.LGG$results[[1]], tissue.definition == "Solid Tissue Normal")) # 0 Normal samples
nrow(subset(query.maskCNV.hg38.LGG$results[[1]], tissue.definition == "Primary solid Tumor")) # 514 Tumor samples
nrow(subset(query.maskCNV.hg38.LGG$results[[1]], tissue.definition == "Blood Derived Normal")) # 485 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.LGG)
maskCNV.LGG <- GDCprepare(query = query.maskCNV.hg38.LGG)
save(maskCNV.LGG,
     file = "../data/maskCNV.LGG.RData")

## TCGA-HNSC Masked Copy Number Variations
query.maskCNV.hg38.HNSC <- GDCquery(project = "TCGA-HNSC",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.HNSC$results[[1]]) # Total number of samples = Tumor + Normal = 1083
nrow(subset(query.maskCNV.hg38.HNSC$results[[1]], tissue.definition == "Solid Tissue Normal")) # 74 Normal samples
nrow(subset(query.maskCNV.hg38.HNSC$results[[1]], tissue.definition == "Primary solid Tumor")) # 519 Tumor samples
nrow(subset(query.maskCNV.hg38.HNSC$results[[1]], tissue.definition == "Blood Derived Normal")) # 488 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.HNSC)
maskCNV.HNSC <- GDCprepare(query = query.maskCNV.hg38.HNSC)
save(maskCNV.HNSC,
     file = "../data/maskCNV.HNSC.RData")

## TCGA-TGCT Masked Copy Number Variations
query.maskCNV.hg38.TGCT <- GDCquery(project = "TCGA-TGCT",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.TGCT$results[[1]]) # Total number of samples = Tumor + Normal = 271
nrow(subset(query.maskCNV.hg38.TGCT$results[[1]], tissue.definition == "Solid Tissue Normal")) # 0 Normal samples
nrow(subset(query.maskCNV.hg38.TGCT$results[[1]], tissue.definition == "Primary solid Tumor")) # 134 Tumor samples
nrow(subset(query.maskCNV.hg38.TGCT$results[[1]], tissue.definition == "Blood Derived Normal")) # 132 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.TGCT)
maskCNV.TGCT <- GDCprepare(query = query.maskCNV.hg38.TGCT)
save(maskCNV.TGCT,
     file = "../data/maskCNV.TGCT.RData")

## TCGA-THCA Masked Copy Number Variations
query.maskCNV.hg38.THCA <- GDCquery(project = "TCGA-THCA",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.THCA$results[[1]]) # Total number of samples = Tumor + Normal = 1025
nrow(subset(query.maskCNV.hg38.THCA$results[[1]], tissue.definition == "Solid Tissue Normal")) # 96 Normal samples
nrow(subset(query.maskCNV.hg38.THCA$results[[1]], tissue.definition == "Primary solid Tumor")) # 505 Tumor samples
nrow(subset(query.maskCNV.hg38.THCA$results[[1]], tissue.definition == "Blood Derived Normal")) # 417 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.THCA)
maskCNV.THCA <- GDCprepare(query = query.maskCNV.hg38.THCA)
save(maskCNV.THCA,
     file = "../data/maskCNV.THCA.RData")

## TCGA-LIHC Masked Copy Number Variations
query.maskCNV.hg38.LIHC <- GDCquery(project = "TCGA-LIHC",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.LIHC$results[[1]]) # Total number of samples = Tumor + Normal = 768
nrow(subset(query.maskCNV.hg38.LIHC$results[[1]], tissue.definition == "Solid Tissue Normal")) # 83 Normal samples
nrow(subset(query.maskCNV.hg38.LIHC$results[[1]], tissue.definition == "Primary solid Tumor")) # 376 Tumor samples
nrow(subset(query.maskCNV.hg38.LIHC$results[[1]], tissue.definition == "Blood Derived Normal")) # 306 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.LIHC)
maskCNV.LIHC <- GDCprepare(query = query.maskCNV.hg38.LIHC)
save(maskCNV.LIHC,
     file = "../data/maskCNV.LIHC.RData")

## TCGA-BLCA Masked Copy Number Variations
query.maskCNV.hg38.BLCA <- GDCquery(project = "TCGA-BLCA",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.BLCA$results[[1]]) # Total number of samples = Tumor + Normal = 814
nrow(subset(query.maskCNV.hg38.BLCA$results[[1]], tissue.definition == "Solid Tissue Normal")) # 34 Normal samples
nrow(subset(query.maskCNV.hg38.BLCA$results[[1]], tissue.definition == "Primary solid Tumor")) # 416 Tumor samples
nrow(subset(query.maskCNV.hg38.BLCA$results[[1]], tissue.definition == "Blood Derived Normal")) # 363 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.BLCA)
maskCNV.BLCA <- GDCprepare(query = query.maskCNV.hg38.BLCA)
save(maskCNV.BLCA,
     file = "../data/maskCNV.BLCA.RData")

## TCGA-UCEC Masked Copy Number Variations
query.maskCNV.hg38.UCEC <- GDCquery(project = "TCGA-UCEC",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.UCEC$results[[1]]) # Total number of samples = Tumor + Normal = 1094
nrow(subset(query.maskCNV.hg38.UCEC$results[[1]], tissue.definition == "Solid Tissue Normal")) # 38 Normal samples
nrow(subset(query.maskCNV.hg38.UCEC$results[[1]], tissue.definition == "Primary solid Tumor")) # 548 Tumor samples
nrow(subset(query.maskCNV.hg38.UCEC$results[[1]], tissue.definition == "Blood Derived Normal")) # 507 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.UCEC)
maskCNV.UCEC <- GDCprepare(query = query.maskCNV.hg38.UCEC)
save(maskCNV.UCEC,
     file = "../data/maskCNV.UCEC.RData")

## TCGA-PRAD Masked Copy Number Variations
query.maskCNV.hg38.PRAD <- GDCquery(project = "TCGA-PRAD",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.PRAD$results[[1]]) # Total number of samples = Tumor + Normal = 1038
nrow(subset(query.maskCNV.hg38.PRAD$results[[1]], tissue.definition == "Solid Tissue Normal")) # 113 Normal samples
nrow(subset(query.maskCNV.hg38.PRAD$results[[1]], tissue.definition == "Primary solid Tumor")) # 501 Tumor samples
nrow(subset(query.maskCNV.hg38.PRAD$results[[1]], tissue.definition == "Blood Derived Normal")) # 423 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.PRAD)
maskCNV.PRAD <- GDCprepare(query = query.maskCNV.hg38.PRAD)
save(maskCNV.PRAD,
     file = "../data/maskCNV.PRAD.RData")

## TCGA-OV Masked Copy Number Variations
query.maskCNV.hg38.OV <- GDCquery(project = "TCGA-OV",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.OV$results[[1]]) # Total number of samples = Tumor + Normal = 1146
nrow(subset(query.maskCNV.hg38.OV$results[[1]], tissue.definition == "Solid Tissue Normal")) # 121 Normal samples
nrow(subset(query.maskCNV.hg38.OV$results[[1]], tissue.definition == "Primary solid Tumor")) # 573 Tumor samples
nrow(subset(query.maskCNV.hg38.OV$results[[1]], tissue.definition == "Blood Derived Normal")) # 425 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.OV)
maskCNV.OV <- GDCprepare(query = query.maskCNV.hg38.OV)
save(maskCNV.OV,
     file = "../data/maskCNV.OV.RData")

## TCGA-THYM Masked Copy Number Variations
query.maskCNV.hg38.THYM <- GDCquery(project = "TCGA-THYM",
             data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.THYM$results[[1]]) # Total number of samples = Tumor + Normal = 249
nrow(subset(query.maskCNV.hg38.THYM$results[[1]], tissue.definition == "Solid Tissue Normal")) # 12 Normal samples
nrow(subset(query.maskCNV.hg38.THYM$results[[1]], tissue.definition == "Primary solid Tumor")) # 124 Tumor samples
nrow(subset(query.maskCNV.hg38.THYM$results[[1]], tissue.definition == "Blood Derived Normal")) # 111 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.THYM)
maskCNV.THYM <- GDCprepare(query = query.maskCNV.hg38.THYM)
save(maskCNV.THYM,
     file = "../data/maskCNV.THYM.RData")

## TCGA-CHOL Masked Copy Number Variations
query.maskCNV.hg38.CHOL <- GDCquery(project = "TCGA-CHOL",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.CHOL$results[[1]]) # Total number of samples = Tumor + Normal = 85
nrow(subset(query.maskCNV.hg38.CHOL$results[[1]], tissue.definition == "Solid Tissue Normal")) # 15 Normal samples
nrow(subset(query.maskCNV.hg38.CHOL$results[[1]], tissue.definition == "Primary solid Tumor")) # 36 Tumor samples
nrow(subset(query.maskCNV.hg38.CHOL$results[[1]], tissue.definition == "Blood Derived Normal")) # 34 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.CHOL)
maskCNV.CHOL <- GDCprepare(query = query.maskCNV.hg38.CHOL)
save(maskCNV.CHOL,
     file = "../data/maskCNV.CHOL.RData")

## TCGA-STAD Masked Copy Number Variations
query.maskCNV.hg38.STAD <- GDCquery(project = "TCGA-STAD",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.STAD$results[[1]]) # Total number of samples = Tumor + Normal = 906
nrow(subset(query.maskCNV.hg38.STAD$results[[1]], tissue.definition == "Solid Tissue Normal")) # 95 Normal samples
nrow(subset(query.maskCNV.hg38.STAD$results[[1]], tissue.definition == "Primary solid Tumor")) # 442 Tumor samples
nrow(subset(query.maskCNV.hg38.STAD$results[[1]], tissue.definition == "Blood Derived Normal")) # 369 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.STAD)
maskCNV.STAD <- GDCprepare(query = query.maskCNV.hg38.STAD)
save(maskCNV.STAD,
     file = "../data/maskCNV.STAD.RData")

## TCGA-LUAD Masked Copy Number Variations
query.maskCNV.hg38.LUAD <- GDCquery(project = "TCGA-LUAD",
              data.category = "Copy Number Variation",
              data.type = "Masked Copy Number Segment")
# Number of samples
nrow(query.maskCNV.hg38.LUAD$results[[1]]) # Total number of samples = Tumor + Normal = 1147
nrow(subset(query.maskCNV.hg38.LUAD$results[[1]], tissue.definition == "Solid Tissue Normal")) # 178 Normal samples
nrow(subset(query.maskCNV.hg38.LUAD$results[[1]], tissue.definition == "Primary solid Tumor")) # 554 Tumor samples
nrow(subset(query.maskCNV.hg38.LUAD$results[[1]], tissue.definition == "Blood Derived Normal")) # 413 Blood-derived Normal samples

GDCdownload(query = query.maskCNV.hg38.LUAD)
maskCNV.LUAD <- GDCprepare(query = query.maskCNV.hg38.LUAD)
save(maskCNV.LUAD,
     file = "../data/maskCNV.LUAD.RData")


##### Combining all maskCNV data into one data frame #####
tcgaProjects <- c("ACC", "BLCA", "BRCA", "CESC", "CHOL", "COAD" ,"DLBC", "ESCA", "GBM", "HNSC", "KICH", "KIRC" ,"KIRP", "LAML", "LGG", "LIHC", "LUAD", "LUSC",
                  "MESO", "OV", "PAAD", "PCPG", "PRAD", "READ", "SARC", "SKCM", "STAD", "TGCT", "THCA", "THYM", "UCEC", "UCS", "UVM")
# Adding on the tumor type to the data frame
for (cancer in 1:length(tcgaProjects)) {
  assign(paste("maskCNV.", tcgaProjects[cancer], sep = ""),
         cbind(get(paste("maskCNV.", tcgaProjects[cancer], sep = "")),
               tcgaProjects[cancer]))
}


# Changing the column name for the tumor types
colnames(maskCNV.ACC) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.BLCA) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.BRCA) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.CESC) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.CHOL) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.COAD) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.DLBC) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.ESCA) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.GBM) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.HNSC) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.KICH) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.KIRC) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.KIRP) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.LAML) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.LGG) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.LIHC) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.LUAD) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.LUSC) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.MESO) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.OV) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.PAAD) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.PCPG) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.PRAD) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.READ) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.SARC) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.SKCM) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.STAD) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.TGCT) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.THCA) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.THYM) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.UCEC) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.UCS) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")
colnames(maskCNV.UVM) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Project")

# Binding all the rows together 
maskCNV.TCGA <- rbind(maskCNV.ACC, maskCNV.BLCA, maskCNV.BRCA, maskCNV.CESC, maskCNV.CHOL, maskCNV.COAD, maskCNV.DLBC, maskCNV.ESCA, maskCNV.GBM, maskCNV.HNSC, maskCNV.KICH,
                      maskCNV.KIRC, maskCNV.KIRP, maskCNV.LAML, maskCNV.LGG, maskCNV.LIHC, maskCNV.LUAD, maskCNV.LUSC, maskCNV.MESO, maskCNV.OV, maskCNV.PAAD, maskCNV.PCPG,
                      maskCNV.PRAD, maskCNV.READ, maskCNV.SARC, maskCNV.SKCM, maskCNV.STAD, maskCNV.TGCT, maskCNV.THCA, maskCNV.THYM, maskCNV.UCEC, maskCNV.UCS, maskCNV.UVM)

# Splitting out Patient ID and adding it to the maskCNV.TCGA data frame
patientid <- stringr::str_match(string = maskCNV.TCGA$Sample,
                                pattern = "[[:upper:]]{4}[-][[:alnum:]]{2}[-]([[:alnum:]]{4}[-][[:alnum:]]{2})[[:alnum:]][-][[:print:]]+")
patientid <- as.data.frame(patientid)
patientid_2 <- as.vector(patientid[,2])

# Combine
maskCNV.TCGA <- data.table(maskCNV.TCGA$GDC_Aliquot, maskCNV.TCGA$Chromosome, maskCNV.TCGA$Start, maskCNV.TCGA$End, maskCNV.TCGA$Num_Probes, maskCNV.TCGA$Segment_Mean,
                           maskCNV.TCGA$Sample, patientid_2, maskCNV.TCGA$Project)
colnames(maskCNV.TCGA) <- c("GDC_Aliquot", "Chromosome", "Start", "End", "Num_Probes", "Segment_Mean", "Sample", "Patient ID", "Project")
# Fix all the data types
maskCNV.TCGA$Chromosome <- as.numeric(maskCNV.TCGA$Chromosome)
maskCNV.TCGA$Project <- as.character(maskCNV.TCGA$Project)

# Getting Tumor IDs for all projects
tumorIDs.TCGA <- TCGAquery_SampleTypes(maskCNV.TCGA$Sample,
                                       typesample = "TP")
tumorBloodIDs.TCGA <- TCGAquery_SampleTypes(maskCNV.TCGA$Sample,
                                            typesample = "TB")
metastaticIDs.TCGA <- TCGAquery_SampleTypes(maskCNV.TCGA$Sample,
                                            typesample = "TM")

# Getting Normal IDs for all projects
normalIDs.TCGA <- TCGAquery_SampleTypes(maskCNV.TCGA$Sample,
                                        typesample = "NT")
bloodIDs.TCGA <- TCGAquery_SampleTypes(maskCNV.TCGA$Sample,
                                       typesample = "NB")

# Saveeeeeeee
save(maskCNV.TCGA, tumorIDs.TCGA, tumorBloodIDs.TCGA, metastaticIDs.TCGA, normalIDs.TCGA, bloodIDs.TCGA,
     file = "../data/maskCNV.TCGA.RData")

##### Clinical Data #####
# Clinical Data for all TCGA Projects 
clinical.KIRC <- GDCquery_clinic(project = "TCGA-KIRC")
clinical.BRCA <- GDCquery_clinic(project = "TCGA-BRCA")
clinical.SARC <- GDCquery_clinic(project = "TCGA-SARC")
clinical.MESO <- GDCquery_clinic(project = "TCGA-MESO")
clinical.READ <- GDCquery_clinic(project = "TCGA-READ")
clinical.KIRP <- GDCquery_clinic(project = "TCGA-KIRP")
clinical.PAAD <- GDCquery_clinic(project = "TCGA-PAAD")
clinical.GBM <- GDCquery_clinic(project = "TCGA-GBM")
clinical.ACC <- GDCquery_clinic(project = "TCGA-ACC")
clinical.CESC <- GDCquery_clinic(project = "TCGA-CESC")
clinical.ESCA <- GDCquery_clinic(project = "TCGA-ESCA")
clinical.DLBC <- GDCquery_clinic(project = "TCGA-DLBC")
clinical.KICH <- GDCquery_clinic(project = "TCGA-KICH")
clinical.UVM <- GDCquery_clinic(project = "TCGA-UVM")
clinical.LAML <- GDCquery_clinic(project = "TCGA-LAML")
clinical.SKCM <- GDCquery_clinic(project = "TCGA-SKCM")
clinical.PCPG <- GDCquery_clinic(project = "TCGA-PCPG")
clinical.COAD <- GDCquery_clinic(project = "TCGA-COAD")
clinical.UCS <- GDCquery_clinic(project = "TCGA-UCS")
clinical.LUSC <- GDCquery_clinic(project = "TCGA-LUSC")
clinical.LGG <- GDCquery_clinic(project = "TCGA-LGG")
clinical.HNSC <- GDCquery_clinic(project = "TCGA-HNSC")
clinical.TGCT <- GDCquery_clinic(project = "TCGA-TGCT")
clinical.THCA <- GDCquery_clinic(project = "TCGA-THCA")
clinical.LIHC <- GDCquery_clinic(project = "TCGA-LIHC")
clinical.BLCA <- GDCquery_clinic(project = "TCGA-BLCA")
clinical.UCEC <- GDCquery_clinic(project = "TCGA-UCEC")
clinical.PRAD <- GDCquery_clinic(project = "TCGA-PRAD")
clinical.OV <- GDCquery_clinic(project = "TCGA-OV")
clinical.THYM <- GDCquery_clinic(project = "TCGA-THYM")
clinical.CHOL <- GDCquery_clinic(project = "TCGA-CHOL")
clinical.STAD <- GDCquery_clinic(project = "TCGA-STAD")
clinical.LUAD <- GDCquery_clinic(project = "TCGA-LUAD")

# Combining it all into a single data frame
clinical.tcga <- rbind(clinical.ACC, clinical.BLCA, clinical.BRCA, clinical.CESC, clinical.CHOL, clinical.COAD, clinical.DLBC, clinical.ESCA, clinical.GBM, clinical.HNSC, clinical.KICH,
                       clinical.KIRC, clinical.KIRP, clinical.LAML, clinical.LGG, clinical.LIHC, clinical.LUAD, clinical.LUSC, clinical.MESO, clinical.OV, clinical.PAAD, clinical.PCPG,
                       clinical.PRAD, clinical.READ, clinical.SARC, clinical.SKCM, clinical.STAD, clinical.TGCT, clinical.THCA, clinical.THYM, clinical.UCEC, clinical.UCS, clinical.UVM)

```


Gene expression
```{r}
### Title: Get all Gene expression data from TCGA
### Written by: Roshan Darji
### Date: 2018.11.19
### Last Updated: 2018.11.20
### Goal: Obtain and combine all TCGA gene expression data into one object

##### Packages #####
suppressPackageStartupMessages(library(TCGAbiolinks))
suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(tidyverse))


##### Obtaining TCGA Gene Expression Data #####
# Creating a vector with all the TCGA Cancer acronyms 
tcgaCancerList <- c("ACC", "BLCA", "BRCA", "CESC", "CHOL", "COAD", "DLBC", "ESCA", "GBM", "HNSC", "KICH",
                    "KIRC", "KIRP", "LAML", "LGG", "LIHC", "LUAD", "LUSC", "MESO", "OV", "PAAD", "PCPG",
                    "PRAD", "READ", "SARC", "SKCM", "STAD", "TGCT", "THCA", "THYM", "UCEC", "UCS", "UVM")
# Looping to query all TCGA projects
setwd("general/")
for (cancer in 1:length(tcgaCancerList)) {
  assign(paste("query.exp.hg38.", tcgaCancerList[cancer], sep = ""),
         GDCquery(project = paste("TCGA-", tcgaCancerList[cancer], sep = ""),
                  data.category = "Transcriptome Profiling",
                  data.type = "Gene Expression Quantification",
                  workflow.type = "HTSeq - Counts"))
}
beepr::beep(sound = 5)

# Looping to download all TCGA projects
tcgaQueryList <- ls(pattern = "query.exp.hg38.")
for (query in 1:length(tcgaQueryList)) {
  GDCdownload(query = base::get(tcgaQueryList[query]))
}

# Looping to prepare all TCGA projects
for (query in 1:length(tcgaQueryList)) {
  assign(paste("exp.", tcgaCancerList[query], sep = ""),
         GDCprepare(query = base::get(tcgaQueryList[query])))
 }
beepr::beep(sound = 5)

##### Combining all data into one object #####
tcgaExpList <- ls(pattern = "^exp.")
exp.tcga.list <- list()
for (cancer in 1:length(tcgaExpList)) {
  exp.tcga.list[[cancer]] <- get(tcgaExpList[cancer])
  exp.tcga.list[[cancer]] <- exp.tcga.list[[cancer]][, !duplicated(exp.tcga.list[[cancer]]@colData$sample)] # Removes duplicates
}
names(exp.tcga.list) <- tcgaCancerList
rm(list = c(tcgaExpList, tcgaQueryList))

##### Combinded RangedSummarizedExperiment object instead of list #####
exp.tcga <- exp.tcga.list[[1]]
for (cancer in 2:length(exp.tcga.list)) {
  exp.tcga <- cbind(exp.tcga, exp.tcga.list[[cancer]])
}

##### Saving the data #####
setwd("../")
save(exp.tcga,
     file = "general/get-tcga/data/190205_getTCGA-exp_exp.tcga.RData")
save(exp.tcga.list,
     file = "general/get-tcga/data/190220_getTCGA-exp_exp.tcga.list.RData")
```


----------------------------------------------------------------------------------------

```{r}
library(TCGAbiolinks)
library(SummarizedExperiment)
library(dplyr)
library(tidyr)
library(stringr)
library(here)
```


## maskCNV data
```{r}

## We create a vector with 33 cancer names to loop through
cancer.project <- c("TCGA-KIRC", "TCGA-BRCA", "TCGA-SARC", "TCGA-MESO", "TCGA-READ", "TCGA-KIRP", "TCGA-PAAD", "TCGA-GBM", "TCGA-ACC", "TCGA-CESC", "TCGA-ESCA", "TCGA-DLBC", "TCGA-KICH", "TCGA-UVM", "TCGA-LAML", "TCGA-SKCM", "TCGA-PCPG", "TCGA-COAD", "TCGA-UCS", "TCGA-LUSC", "TCGA-LGG", "TCGA-HNSC", "TCGA-TGCT", "TCGA-THCA", "TCGA-LIHC", "TCGA-BLCA", "TCGA-UCEC", "TCGA-PRAD", "TCGA-OV", "TCGA-THYM", "TCGA-CHOL", "TCGA-STAD", "TCGA-LUAD")

## The function takes cancer project name as defined in TCGAbiolinks package and returns maskCNV data ## merged with selected clinical variable data as well few extra variables (patient, tumor, project, ## type) 
get_cnv_data <- function(cancer.project){
  query.cancer <- GDCquery(project = cancer.project,
              data.category = "Copy Number Variation",
             data.type = "Masked Copy Number Segment", legacy=FALSE)
  query.result <- query.cancer$results[[1]] %>% select(cases, project, sample_type)
  GDCdownload(query = query.cancer)
  maskCNV.cancer <- GDCprepare(query = query.cancer, summarizedExperiment = TRUE)
  maskCNV.cancer <- inner_join(maskCNV.cancer, query.result, by = c("Sample" = "cases"), copy = TRUE)
  maskCNV.cancer <- maskCNV.cancer %>% separate(Sample, c("patient", "type"), 12, remove = FALSE) %>%                     separate(type, c("tumour"), -13)
  maskCNV.cancer$tumour <- as.numeric(str_remove(maskCNV.cancer$tumour, "-"))
  maskCNV.cancer <- maskCNV.cancer %>% mutate(type = case_when(tumour == 1 | tumour == 6 ~ "tumor",
                    tumour == 10 | tumour == 11 ~ "normal"))
  query.clinic <- GDCquery_clinic(project = cancer.project) %>% select(any_of(c("submitter_id", 
                  "ajcc_pathologic_t", "ajcc_pathologic_m", "ajcc_pathologic_n", "ajcc_clinical_m", 
                  "age_at_diagnosis", "gender", "ajcc_pathologic_stage")))
  query.map <- inner_join(maskCNV.cancer, query.clinic, by = c("patient" = "submitter_id"))
  return(query.map)
}

## We create an empty data frame and loop through all the cancer project and store the output of
## get_cnv_data function in a single data frame and store it in a csv file
all.cancer.df <- data.frame(stringsAsFactors = FALSE)
for (i in 1:length(cancer.project)){
  all.cancer.df <- bind_rows(all.cancer.df, get_cnv_data(cancer.project[i]))
}
write.csv(all.cancer.df, file = here("data", "processed", "all_cancer_df.csv"), row.names = FALSE)

#all.cancer.list <- lapply(cancer.project, function(x) get_cnv_data(x))
```
